R/add-subvariable.R

```diff
@@ -12,25 +13,55 @@
 #' }
 #' @export
 addSubvariable <- function (variable, subvariable) {
-    ## Get subvariable URL or URLs, depending on how many supplied
-    new.urls <- urls(subvariable)
+    subvariable <- add_subvar_def(variable, subvariable)
 
-    ## Store these for post workaround
-    subvar.urls <- subvariables(tuple(variable))
+    ## There is some inconsistency in the output order after patching the subvariable catalog
+    ## this stores the proper order.
+    order <- c(names(subvariables(variable)), vapply(subvariable, name, ""))
 
+    ## Get subvariable URL or URLs, depending on how many supplied
+    new.urls <- urls(subvariable)
     ## Do the adding
     crPATCH(shojiURL(variable, "catalogs", "subvariables"),
         body=toJSON(sapply(new.urls, emptyObject, simplify=FALSE)))
-
-    ## Workaround because apparently bind/rebind isn't retaining the order
-    crPATCH(self(variable),
-        body=toJSON(list(subvariables=I(c(subvar.urls, new.urls)))))
-
+    variable <- refresh(variable)
+    subvariables(variable) <- subvariables(variable)[order]
     ## Refresh and return
     dropCache(datasetReference(variable))
-    invisible(refresh(variable))
+    return(invisible(refresh(variable)))
 }
 
 #' @rdname addSubvariable
 #' @export
 addSubvariables <- addSubvariable
+
+add_subvar_def <- function(var, subvar){
+    ## Input can be a variable, subvariable, dataset subset or
+    ## a mixed or uniform list of variables and subvariables this
+    ## wraps single entries in a list for type consistency.
+    if (inherits(subvar, "VariableDefinition") ||
+            is.variable(subvar)) {
+        ## wrap single variables in list
+        subvar <- list(subvar)
+    }
+
+    vardefs <- vapply(subvar,
+        function(x) inherits(x, "VariableDefinition"),
+        logical(1))
+
+    if (any(vardefs)) {
+        ds <- loadDataset(datasetReference(var))
+        ds <- addVariables(ds, subvar[vardefs])
+
+        ## Since this function accepts mixed lists of variables and definitions
+        ## we need to replace the definitions with the variables once they've been added.
+        subvar[vardefs] <- lapply(subvar[vardefs], function(x){
+            if (is.variable(x)) {
+                x
+            } else {
+                ds[[x$name]]
```

Comment:
Like I said above, this statement shouldn't be necessary, but for reference, this isn't correct. Variables are indexed by alias by default in datasets, so this will only work if the name and alias are the same, or if the user has changed that default. See "namekey" elsewhere in the code.